### PR TITLE
CXX-796 Improve bsoncxx doc comments

### PR DESCRIPTION
**Pull request by hanumantmk**
Replaced doc todo's under bsoncxx with doc comments.

### CODE REVIEW

File: src/bsoncxx/array/element.hpp [hunk 1]
```diff
@@ -27,7 +27,13 @@ BSONCXX_INLINE_NAMESPACE_BEGIN
...
 
 ///
-/// @todo document this class
+/// A variant view type that accesses values in serialized BSON arrays.
+///
+/// Element functions as a variant type, where the kind of the element can be
+/// interrogated by calling type() and a specific value can be extracted through
+/// get_X() accessors.
+///
+/// @relatesalso document::element
```
Comment (samantharitter) [inline src/bsoncxx/array/element.hpp]: what does @relatesalso do here?

File: src/bsoncxx/builder/stream/array_context.hpp [hunk 2]
```diff
@@ -34,26 +34,44 @@ class key_context;
...
+    ///   builder.
+    ///
+    /// @param t
+    ///   The value to append
     ///
     template <class T>
     BSONCXX_INLINE typename std::enable_if<
         !(util::is_functor<T, void(array_context<>)>::value ||
           util::is_functor<T, void(single_context)>::value ||
-          std::is_same<T, const close_document_type>::value ||
```
Comment (acmorrow) [inline src/bsoncxx/builder/stream/array_context.hpp]: This doesn't look like a comment change

File: src/bsoncxx/builder/stream/array_context.hpp [hunk 3]
```diff
@@ -62,7 +80,12 @@ class array_context {
     }
 
     ///
-    /// @todo document this method
+    /// << operator for accepting a callable of the form void(array_context) or
```
Comment (acmorrow) [inline src/bsoncxx/builder/stream/array_context.hpp]: Is 'put to operator' preferable to '<< operator'

File: src/bsoncxx/document/element.hpp [hunk 4]
```diff
@@ -59,168 +59,285 @@ class element;
...
+    ///
+    /// @throws bsoncxx::exception if this element is not a b_maxkey.
+    ///
+    /// @return the element's value.
     ///
     types::b_maxkey get_maxkey() const;
 
     ///
-    /// @todo document this method
+    /// Getter for a type
```
Comment (acmorrow) [inline src/bsoncxx/document/element.hpp]: 'the current type'

File: src/bsoncxx/json.cpp [hunk 5]
```diff
@@ -109,7 +109,7 @@ class json_visitor {
     void visit_value(const types::b_oid& value) {
         out << "{" << std::endl;
         pad(1);
-        out << "\"$oid\" : \"" << value.value << "\"" << std::endl;
+        out << "\"$oid\" : \"" << value.value.to_string() << "\"" << std::endl;
```
Comment (acmorrow) [inline src/bsoncxx/json.cpp]: Unrelated?

File: src/bsoncxx/oid.cpp [hunk 6]
```diff
@@ -120,17 +120,5 @@ bool operator!=(const oid& lhs, const oid& rhs) {
     return oid_compare(lhs, rhs) != 0;
 }
 
-std::ostream& operator<<(std::ostream& out, const oid& rhs) {
```
Comment (acmorrow) [inline src/bsoncxx/oid.cpp]: Can you remove the ostream header from above?

File: src/bsoncxx/oid.hpp [hunk 7]
```diff
@@ -14,6 +14,7 @@
 
 #pragma once
 
+#include <array>
 #include <iosfwd>
```
Comment (acmorrow) [inline src/bsoncxx/oid.hpp]: Please remove `<iosfwd>`

File: src/bsoncxx/oid.hpp [hunk 8]
```diff
@@ -114,12 +117,9 @@ class BSONCXX_API oid {
...
-    /// @todo document this method. Also, do we really want this? This is the only place we use ostreams.
-    ///
-    friend BSONCXX_API std::ostream& BSONCXX_CALL operator<<(std::ostream& out, const oid& rhs);
-
+    /// An accessor for the internal data buffer in the oid.
     ///
-    /// @todo document this method
+    /// @return A pointer to the internal buffer holding the oid bytes.
     ///
     const char* bytes() const;
```
Comment (acmorrow) [inline src/bsoncxx/oid.hpp]: With std::array as the implementation, should this just return `const std::array<...>&

File: src/bsoncxx/builder/stream/single_context.hpp [hunk 9]
```diff
@@ -79,11 +82,19 @@ class single_context {
     }
 
    private:
+    BSONCXX_INLINE array_context<single_context> wrap_array() {
```
Comment (samantharitter) [inline src/bsoncxx/builder/stream/single_context.hpp]: Can these still have some documentation, even though they are private?

File: src/bsoncxx/builder/stream/value_context.hpp [hunk 10]
```diff
@@ -28,20 +28,41 @@ namespace builder {
 namespace stream {
 
 ///
-/// @todo document this class
+/// A stream context which expects a value, which will later be followed by
```
Comment (samantharitter) [inline src/bsoncxx/builder/stream/value_context.hpp]: which "can" later be followed?

File: src/bsoncxx/array/element.hpp (hunk 1 above)
Comment (samantharitter) [inline src/bsoncxx/array/element.hpp]: I think I understand this now, nevermind

Comment (samantharitter) [inline src/bsoncxx/array/element.hpp]: Actually maybe I still don't understand, isn't this tag meant to relate functions to classes, and not to relate classes to other classes? https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdrelatesalso

File: src/bsoncxx/types.hpp [hunk 11]
```diff
@@ -85,7 +95,7 @@ struct BSONCXX_API b_double {
 };
 
 ///
-/// @todo document this method
+/// free function comparator for b_double
```
Comment (samantharitter) [inline src/bsoncxx/types.hpp]: @relatesalso b_double, or will that not work on structs?

File: src/bsoncxx/types.hpp [hunk 12]
```diff
@@ -116,7 +129,7 @@ struct BSONCXX_API b_utf8 {
 };
 
 ///
-/// @todo document this method
+/// free function comparator for b_utf8
```
Comment (samantharitter) [inline src/bsoncxx/types.hpp]: same question about @relatesalso

File: src/bsoncxx/types.hpp [hunk 13]
```diff
@@ -377,7 +402,10 @@ struct BSONCXX_API b_code {
     static constexpr auto type_id = type::k_code;
 
     ///
-    /// @todo document this method
+    /// Constructor for b_code.
+    ///
+    /// @param code
+    ///   The js code.
```
Comment (samantharitter) [inline src/bsoncxx/types.hpp]: nit no period here, since the other params don't have them

File: src/bsoncxx/view_or_value.hpp [hunk 14]
```diff
@@ -128,7 +128,9 @@ class view_or_value {
     }
 
     ///
-    /// @todo : document this method
+    /// Get a view for the type.
```
Comment (samantharitter) [inline src/bsoncxx/view_or_value.hpp]: nit, View

**Comment by samantharitter** [comment]
Some @relatesalso questions and nits, otherwise LGTM

File: src/bsoncxx/builder/stream/single_context.hpp (hunk 9 above)
Comment (acmorrow) [inline src/bsoncxx/builder/stream/single_context.hpp]: They can, but it shouldn't be `/// ` comments, just plain old `//` comments

File: src/bsoncxx/json.cpp (hunk 5 above)
Comment (acmorrow) [inline src/bsoncxx/json.cpp]: oh, I see. nm

File: src/bsoncxx/types.hpp (hunk 11 above)
Comment (acmorrow) [inline src/bsoncxx/types.hpp]: also, @relates or @relatesalso?

File: src/bsoncxx/builder/stream/array_context.hpp (hunk 2 above)
Comment (hanumantmk) [inline src/bsoncxx/builder/stream/array_context.hpp]: we don't ever close documents from array_context.  It's just dead code

Comment (acmorrow) [inline src/bsoncxx/builder/stream/array_context.hpp]: ack

File: src/bsoncxx/builder/stream/array_context.hpp (hunk 3 above)
Comment (hanumantmk) [inline src/bsoncxx/builder/stream/array_context.hpp]: I think it's just as well

File: src/bsoncxx/builder/stream/single_context.hpp (hunk 9 above)
Comment (hanumantmk) [inline src/bsoncxx/builder/stream/single_context.hpp]: I'd just as soon leave them out.  There's not much value to a maintainer or implementer beyond what the class comment indicates, and as private methods they don't need generated docs

File: src/bsoncxx/builder/stream/value_context.hpp (hunk 10 above)
Comment (hanumantmk) [inline src/bsoncxx/builder/stream/value_context.hpp]: makes sense

File: src/bsoncxx/document/element.hpp (hunk 4 above)
Comment (hanumantmk) [inline src/bsoncxx/document/element.hpp]: bleh it's for a value.  I'll fix up again

File: src/bsoncxx/oid.cpp (hunk 6 above)
Comment (hanumantmk) [inline src/bsoncxx/oid.cpp]: yep

File: src/bsoncxx/oid.hpp (hunk 7 above)
Comment (hanumantmk) [inline src/bsoncxx/oid.hpp]: done

File: src/bsoncxx/oid.hpp (hunk 8 above)
Comment (hanumantmk) [inline src/bsoncxx/oid.hpp]: That almost feels like an implementation detail.  I think the raw pointer return is okay

File: src/bsoncxx/types.hpp (hunk 13 above)
Comment (hanumantmk) [inline src/bsoncxx/types.hpp]: done

File: src/bsoncxx/view_or_value.hpp (hunk 14 above)
Comment (hanumantmk) [inline src/bsoncxx/view_or_value.hpp]: done

File: src/bsoncxx/array/element.hpp (hunk 1 above)
Comment (hanumantmk) [inline src/bsoncxx/array/element.hpp]: getting rid of relates, doesn't seem like I was using them vaguely correctly

File: src/bsoncxx/types.hpp (hunk 11 above)
Comment (hanumantmk) [inline src/bsoncxx/types.hpp]: done for these

File: src/bsoncxx/types.hpp (hunk 12 above)
Comment (hanumantmk) [inline src/bsoncxx/types.hpp]: done

File: src/bsoncxx/builder/stream/key_context.hpp [hunk 15]
```diff
@@ -99,7 +147,10 @@ class key_context {
     }
 
     ///
-    /// @todo document this method
+    /// << operator for closing a subdocument in the core builder.
+    ///
+    /// @param _
```
Comment (acmorrow) [inline src/bsoncxx/builder/stream/key_context.hpp]: missing param name

Comment (acmorrow) [inline src/bsoncxx/builder/stream/key_context.hpp]: nm

**Comment by acmorrow** [comment]
LGTM

**Comment by acmorrow** [comment]
Can we get this merged? It has two LGTMs

**Comment by acmorrow** [comment]
Merged in 6e3c2fd837a3ab86f91d43f8a3eed3115caa5537
